File: main/src/main/java/net/fwitz/math/main/binary/escape/CollatzPlot.java
package net.fwitz.math.main.binary.escape;

import net.fwitz.math.fractal.escape.EscapeFunction;
import net.fwitz.math.fractal.escape.collatz.Collatz;
import net.fwitz.math.plot.binary.escape.EscapeTimePlot;
import net.fwitz.math.plot.binary.escape.color.EscapeTimeInterpolator;
import net.fwitz.math.plot.renderer.palette.EscapeTimePalette;

public class CollatzPlot {
    private static final double P_MIN = -20;
    private static final double P_MAX = 20;
    private static final double Q_MIN = -20;
    private static final double Q_MAX = 20;
    private static final int ITERS = 1000;
    private static final int BAILOUT_RADIUS = 1 << 8;
    private static final int BAILOUT = BAILOUT_RADIUS * BAILOUT_RADIUS;

    private static final EscapeFunction COLLATZ = EscapeFunction.builder()
            .step((c, z) -> Collatz.collatz(z))
            .escapeTest(z -> z.abs2() > BAILOUT)
            .maxIters(ITERS)
            .build();

    public static void main(String[] args) {
        EscapeTimePlot.complex("Collatz (Escape time)")
                .computeFn(COLLATZ)
                .domainX(P_MIN, P_MAX)
                .domainY(Q_MIN, Q_MAX)
                .colorFn(new EscapeTimeInterpolator(100, BAILOUT_RADIUS, new EscapeTimePalette()))
                .render();
    }
}

File: main/src/main/kotlin/net/fwitz/math/main/binary/escape/CollatzPlot.kt
package net.fwitz.math.main.binary.escape

import net.fwitz.math.binary.complex.Complex
import net.fwitz.math.fractal.escape.EscapeFunction
import net.fwitz.math.fractal.escape.collatz.Collatz.collatz
import net.fwitz.math.plot.binary.escape.EscapeTimePlot
import net.fwitz.math.plot.binary.escape.color.EscapeTimeInterpolator
import net.fwitz.math.plot.renderer.palette.EscapeTimePalette

object CollatzPlot {
    private const val P_MIN = -20.0
    private const val P_MAX = 20.0
    private const val Q_MIN = -20.0
    private const val Q_MAX = 20.0
    private const val ITERS = 1000
    private const val BAILOUT_RADIUS = 1 shl 8
    private const val BAILOUT = BAILOUT_RADIUS * BAILOUT_RADIUS

    private val COLLATZ = EscapeFunction.builder<Complex>()
        .step { _, z -> collatz(z) }
        .escapeTest { it.abs2 > BAILOUT }
        .maxIters(ITERS)
        .build()

    @JvmStatic
    fun main(args: Array<String>) = EscapeTimePlot.complex("Collatz (Escape time)")
        .computeFn(COLLATZ)
        .domainX(P_MIN, P_MAX)
        .domainY(Q_MIN, Q_MAX)
        .colorFn(EscapeTimeInterpolator(100.0, BAILOUT_RADIUS.toDouble(), EscapeTimePalette))
        .render()
}